src/linear_algebra/vandermonde.lean

```diff
@@ -0,0 +1,313 @@
+    apply if_neg,
+    intro hij',
+    change ↑j < ↑i at hij,
+    rw hij' at hij,
+    exact nat.not_succ_lt_self hij },
+  rw det_of_upper_triangular _ h,
+  apply fintype.prod_eq_one,
+  simp
+end
+
+lemma vandermonde_step3_det : (vandermonde_step2 R n a).det = (vandermonde_step3 R n a).det :=
+begin
+  rw [←mul_one (vandermonde_step2 R n a).det, ←vandermonde_step3_op_det R n a, ←det_mul],
+  congr,
+  ext i k,
+  rw mul_apply,
+  by_cases hk : k = 0,
+  { rw hk,
+    have hj : ∀ j ≠ 0, vandermonde_step2 R n a i j * vandermonde_step3_op R n a j 0 = 0,
+    { intros j hj0,
+      dunfold vandermonde_step3_op,
+      rw [if_neg hj0, if_neg, mul_zero],
+      exact (nat.succ_ne_zero ↑j).symm },
+    rw fintype.sum_eq_single 0 hj,
+    simp only [mul_one, vandermonde_step3_op_diagonal, vandermonde_step2, vandermonde_step3],
+    by_cases hi : i = 0,
+    { rw [if_pos hi, if_pos hi],
+      simp only [fin.coe_zero, if_true, eq_self_iff_true, pow_zero] },
+    { rw [if_neg hi, if_neg hi],
+      simp only [geom_sum₂_zero, fin.coe_zero, if_true, eq_self_iff_true] }},
+  { have hk1 : 1 ≤ ↑k := nat.pos_of_ne_zero (@fin.vne_of_ne _ k 0 hk),
+    have hkmp : ↑k - 1 + 1 = ↑k := nat.sub_add_cancel hk1,
+    have hk1' : ↑(k - 1) = ↑k - 1,
+    { cases n,
+      { exact absurd (fin.eq_zero k) hk },
+      { rw [fin.coe_sub_of_le, fin.coe_one],
+        exact hk1 }},
+    have hk1k : ¬(k - 1) = k,
+    { apply fin.ne_of_vne,
+      rw [fin.val_eq_coe, hk1'],
+      exact λ h, k.val.succ_ne_self.symm ((nat.sub_eq_iff_eq_add hk1).mp h) },
+    have hj : ∀ j,
+      j ≠ (k - 1) ∧ (j ≠ k) → vandermonde_step2 R n a i j * vandermonde_step3_op R n a j k = 0,
+    { intros j h2,
+      dunfold vandermonde_step3_op,
+      rw if_neg h2.right,
+      have hjk : ¬↑k = ↑j + 1,
+      { apply mt (nat.sub_eq_iff_eq_add hk1).mpr,
+        rw ←hk1',
+        exact (fin.vne_of_ne h2.left).symm },
+      rw [if_neg hjk, mul_zero] },
+    rw sum_eq_add (k-1) k hk1k hj,
+    { simp only [vandermonde_step2, vandermonde_step3, vandermonde_step3_op],
+      rw [if_neg hk, if_neg hk1k, if_pos rfl, if_neg hk, hk1', hkmp, if_pos rfl],
+      by_cases hi : i = 0,
+      { rw [if_pos hi, if_pos hi, if_pos hi, mul_comm, ←neg_mul_eq_neg_mul,
+          (pow_succ _ _).symm, hkmp, mul_one, add_left_neg] },
+      { rw [if_neg hi, if_neg hi, if_neg hi, mul_one, mul_comm, ←neg_mul_eq_neg_mul],
+        nth_rewrite 1 ←hkmp,
+        rw geom_sum₂_succ_eq,
+        ring }}},
+end
+
+lemma fin_one_sum_fin_congr_equiv_symm_apply_right {j : fin (n + 1)} (h : j ≠ 0) :
+  (fin_sum_fin_equiv.trans (fin_congr (add_comm 1 n))).symm j = sum.inr (j.pred h) :=
+begin
+  simp only [fin_congr_symm, symm_trans_apply],
+  erw fin_congr_apply_mk (add_comm n 1) _ j.2,
+  rw fin_sum_fin_equiv_symm_apply_right,
+  { simp only [],
+    apply fin.eq_of_veq,
+    simp only [fin.val_eq_coe, fin.coe_pred] },
+  { exact nat.pos_of_ne_zero (λ hj, h (fin.eq_of_veq hj)) }
+end
```

Comment:
Are we missing some lemmas here about `fin_sum_fin_equiv` and `fin_congr`? This seems more painful than I'd expect.